File: src/main/java/wraith/fabricaeexnihilo/modules/infested/InfestedHelper.java
package wraith.fabricaeexnihilo.modules.infested;

import net.minecraft.block.Block;
import net.minecraft.block.Blocks;
import net.minecraft.block.LeavesBlock;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.World;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.modules.ModBlocks;
import wraith.fabricaeexnihilo.recipe.util.Lazy;
import wraith.fabricaeexnihilo.util.RegistryUtils;

import java.util.HashMap;

public final class InfestedHelper {
    
    private InfestedHelper() {
    }
    
    private static final Lazy<HashMap<Block, InfestedLeavesBlock>> LEAF_TO_INFESTED = new Lazy<>(() -> {
        HashMap<Block, InfestedLeavesBlock> map = new HashMap<>();
        for (var entry : ModBlocks.INFESTED_LEAVES.values()) {
            map.put(entry.getLeafBlock(), entry);
        }
        return map;
    });
    
    public static ActionResult tryToInfest(World world, BlockPos pos) {
        var originalState = world.getBlockState(pos);
        var infestedBlock = LEAF_TO_INFESTED.get().get(originalState.getBlock());
        if (infestedBlock == null) {
            return ActionResult.PASS;
        }
        
        var newState = ModBlocks.INFESTING_LEAVES
                .getDefaultState()
                .with(LeavesBlock.DISTANCE, originalState.get(LeavesBlock.DISTANCE))
                .with(LeavesBlock.PERSISTENT, originalState.get(LeavesBlock.PERSISTENT));
        
        world.setBlockState(pos, newState);
        
        if (!(world.getBlockEntity(pos) instanceof InfestingLeavesBlockEntity blockEntity)) {
            FabricaeExNihilo.LOGGER.warn("Placed block doesn't have a block entity! (InfestedHelper)");
            return ActionResult.PASS;
        }
        
        blockEntity.setTarget(infestedBlock);
        
        return ActionResult.SUCCESS;
    }
    
    public static void tryToSpreadFrom(World world, BlockPos pos, int tries) {
        for (var attempt = 0; attempt < tries; ++attempt) {
            if ((world.random.nextFloat()) < FabricaeExNihilo.CONFIG.modules.silkworms.spreadChance) {
                var i = world.random.nextInt(3) - 1;
                var j = world.random.nextInt(3) - 1;
                var k = world.random.nextInt(3) - 1;
                var newPos = pos.add(new Vec3i(i, j, k));
                if (World.isValid(newPos)) {
                    tryToInfest(world, newPos);
                }
            }
        }
    }
}

File: src/main/kotlin/exnihilofabrico/modules/infested/InfestedHelper.kt
package exnihilofabrico.modules.infested

import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.modules.ModBlocks
import net.minecraft.block.Block
import net.minecraft.block.LeavesBlock
import net.minecraft.util.ActionResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3i
import net.minecraft.world.World

object InfestedHelper {
    fun tryToInfest(world: World, pos: BlockPos): ActionResult {
        if(world.isClient) return ActionResult.PASS
        val originalState = world.getBlockState(pos)
        if(!LEAF_TO_INFESTED.containsKey(originalState.block))
            return ActionResult.PASS
        val newState = ModBlocks.INFESTING_LEAVES.defaultState
            .with(LeavesBlock.DISTANCE, originalState.get(LeavesBlock.DISTANCE))
            .with(LeavesBlock.PERSISTENT, originalState.get(LeavesBlock.PERSISTENT))
        val infestedBlock = getInfestedLeavesBlock(originalState.block)

        world.setBlockState(pos, newState)

        val blockEntity = world.getBlockEntity(pos) as? InfestingLeavesBlockEntity ?: return ActionResult.PASS

        blockEntity.infestedBlock = infestedBlock
        blockEntity.markDirty()

        return ActionResult.SUCCESS
    }

    fun tryToSpreadFrom(world: World, pos: BlockPos, tries: Int = 1) {
        for(attempt in (0 until tries)) {
            if((world.random.nextFloat()) < ExNihiloFabrico.config.modules.silkworms.spreadChance) {
                val i = world.random.nextInt(3) - 1
                val j = world.random.nextInt(3) - 1
                val k = world.random.nextInt(3) - 1
                val newPos = pos.add(Vec3i(i,j,k))
                if(world.isHeightValidAndBlockLoaded(newPos))
                    tryToInfest(world, newPos)
            }
        }
    }

    private val LEAF_TO_INFESTED = ModBlocks.INFESTED_LEAVES.map { (_,v) -> v.leafBlock to v }.toMap()

    private fun getInfestedLeavesBlock(block: Block): InfestedLeavesBlock {
        if(LEAF_TO_INFESTED.isEmpty())
            ModBlocks.INFESTED_LEAVES.map { (_,v) -> v.leafBlock to v }.toMap()
        return LEAF_TO_INFESTED[block] ?: ModBlocks.INFESTED_LEAVES.values.first()
    }
}